broker/src/main/java/io/camunda/zeebe/broker/system/restore/PartitionRestoreService.java

```diff
@@ -0,0 +1,250 @@
+    LOG.error(
+        "Cannot find the checkpoint record at position {}. Log contains first record: (index = {}, position= {}) last record: (index = {}, position= {}). Restoring from this state can lead to inconsistent state. Aborting restore.",
+        checkpointPosition,
+        firstEntry != null ? firstEntry.index() : -1,
+        firstEntry != null ? firstEntry.asqn() : -1,
+        lastEntry != null ? lastEntry.index() : -1,
+        lastEntry != null ? lastEntry.asqn() : -1);
+    throw new IllegalStateException(
+        "Failed to restore from backup. Cannot find a record at checkpoint position %d in the log."
+            .formatted(checkpointPosition));
+  }
+  // Move contents of restored backup from the temp directory to partition's root data directory.
+  // After this is done, the contents of the data directory follow the expected directory
+  // structure. That is - segments in rootDirectory, snapshot in
+  // rootDirectory/snapshots/<snapshotId>/
+  private Backup moveFilesToDataDirectory(final Backup backup) {
+    moveSegmentFiles(backup);
+    moveSnapshotFiles(backup);
+    return backup;
+  }
+
+  private void moveSegmentFiles(final Backup backup) {
+    LOG.info("Moving journal segment files to {}", rootDirectory);
+    final var segmentFileSet = backup.segments().namedFiles();
+    final var segmentFileNames = segmentFileSet.keySet();
+    segmentFileNames.forEach(
+        name -> copyNamedFileToDirectory(name, segmentFileSet.get(name), rootDirectory));
+  }
+
+  private void moveSnapshotFiles(final Backup backup) {
+    if (backup.descriptor().snapshotId().isEmpty()) {
+      return;
+    }
+
+    final var snapshotStore =
+        FileBasedSnapshotStoreFactory.createRestorableSnapshotStore(
+            partition.dataDirectory().toPath(), partition.id().id(), localBrokerId);
+
+    try {
+      snapshotStore.restore(
+          backup.descriptor().snapshotId().orElseThrow(), backup.snapshot().namedFiles());
+    } catch (final IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void copyNamedFileToDirectory(
+      final String name, final Path source, final Path targetDirectory) {
+    final var targetFilePath = targetDirectory.resolve(name);
+    try {
+      Files.move(source, targetFilePath);
+    } catch (final IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private CompletionStage<Backup> download(
+      final long checkpointId, final Path tempRestoringDirectory) {
+    return findValidBackup(checkpointId)
+        .thenCompose(
+            backup -> {
+              LOG.info("Downloading backup {} to {}", backup, tempRestoringDirectory);
+              return backupStore.restore(backup, tempRestoringDirectory);
+            });
+  }
+
+  private CompletionStage<BackupIdentifier> findValidBackup(final long checkpointId) {
+    LOG.info("Searching for a completed backup with id {}", checkpointId);
+    final var futures =
```

Comment:
:wrench: We will probably want more diagnostic information here in the future in case nothing is found, e.g. what backups were found and their state. But it's kind of hard to tell at this state (for me anyway) so I'm fine with waiting for now unless you have some idea.

Reply:
On error, all backups that are found is already logged. Did you mean something else?

Reply:
Ah, you mean if the given backupId is not available, list other available backupIds? That makes sense. But currently list all backup is not a supported feature in the backup store. We have plans to add it though.

Now we are logging, the status of the backup with the backupId. So the user can see why it is not complete. It might not exist, or it is in a failed state etc.